### PR TITLE
[Merged by Bors] - fix: mdata in cancel_denoms

### PR DESCRIPTION
**Pull request by adomani**
PR inspired by "automated bugs" (#12054).

[Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/automated.20bugs/near/432861117)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/CancelDenoms/Core.lean [hunk 1]
```diff
@@ -328,7 +329,7 @@ open Elab Tactic
 
 def cancelDenominatorsAt (fvar : FVarId) : TacticM Unit := do
   let t ← instantiateMVars (← fvar.getDecl).type
-  let (new, eqPrf) ← CancelDenoms.cancelDenominatorsInType t
+  let (new, eqPrf) ← CancelDenoms.cancelDenominatorsInType t.consumeMData
```
Comment (eric-wieser) [inline Mathlib/Tactic/CancelDenoms/Core.lean:332]: I don't think this is the right fix; I think `findCompLemma` should be moved to `MetaM`, and should do a `whnfR`

Comment (eric-wieser) [reply Mathlib/Tactic/CancelDenoms/Core.lean:332]: (the current version is also tricked by reducible definitions, not just metadata)

Comment (adomani) [reply Mathlib/Tactic/CancelDenoms/Core.lean:332]: Ok, this change means that `≠` does not pass as `Ne`, but as `¬ Eq`: I'll fix this in a bit!

Comment (adomani) [reply Mathlib/Tactic/CancelDenoms/Core.lean]: A suggested by Eric, I pushed a change with `whnfR` in `findCompLemma` and therefore bypassed the need for `consumeMData/cleanupAnnotations`.

File: Mathlib/Tactic/CancelDenoms/Core.lean [hunk 2]
```diff
@@ -259,15 +259,18 @@ def derive (e : Expr) : MetaM (ℕ × Expr) := do
...
-  | (``Ne, #[_, a, b]) => (a, b, ``cancel_factors_ne, false)
-  | (``GE.ge, #[_, _, a, b]) => (b, a, ``cancel_factors_le, true)
-  | (``GT.gt, #[_, _, a, b]) => (b, a, ``cancel_factors_lt, true)
-  | _ => none
+def findCompLemma (e : Expr) : MetaM (Option (Expr × Expr × Name × Bool)) := do
+  match (← whnfR e).getAppFnArgs with
+  | (``LT.lt, #[_, _, a, b]) => return (a, b, ``cancel_factors_lt, true)
+  | (``LE.le, #[_, _, a, b]) => return (a, b, ``cancel_factors_le, true)
+  | (``Eq, #[_, a, b]) => return (a, b, ``cancel_factors_eq, false)
+  -- `a ≠ b` reduces to `¬ a = b` under `whnf`
```
Comment (eric-wieser) [inline Mathlib/Tactic/CancelDenoms/Core.lean:267]: Can you add a test for this? This ought to do:
```
example : ¬ (4 / 2 : ℚ) = 3 := by cancel_denoms; norm_num
```
which previously failed.

**Review by eric-wieser — APPROVED** [review]
bors d+

Thanks!

File: Mathlib/Tactic/CancelDenoms/Core.lean (hunk 2 above)
Comment (adomani) [reply Mathlib/Tactic/CancelDenoms/Core.lean:267]: Added!

**Pull request closed without merging**